In [14]:
import pandas as pd
import seaborn as sns

In [15]:
mpg = sns.load_dataset('mpg')
mpg

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa                 chevy s-10  

[398 rows x 9 columns]

In [4]:
#데이터셋 정보 확인 (데이터의 열 정보, 결측값 여부 등)
mpg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


In [5]:
mpg.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

In [6]:
mpg.head(5)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [7]:
mpg.shape

(398, 9)

In [8]:
mpg.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64

In [10]:
mpg.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model_year        int64
origin           object
name             object
dtype: object

In [22]:
# 1. model_year 기준으로 그룹화하여 각 연도별로 가장 높은 연비(mpg)를 가진 차량 찾기
group_by_year = mpg.groupby('model_year') 
group_by_year

In [18]:
# 2. 각 그룹에서 연비(mpg)가 가장 큰 값 찾기
max_mpg_per_year = group_by_year['mpg'].max()  # 연도별 최대 mpg 추출
max_mpg_per_year

model_year
70    27.0
71    35.0
72    28.0
73    29.0
74    32.0
75    33.0
76    33.0
77    36.0
78    43.1
79    37.3
80    46.6
81    39.1
82    44.0
Name: mpg, dtype: float64

In [24]:
# 3. 가장 높은 연비를 가진 차량의 이름을 찾기 위해 merge
result = pd.merge(mpg, max_mpg_per_year, on=['model_year', 'mpg'])
result

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0   27.0          4          97.0        88.0    2130          14.5   
1   35.0          4          72.0        69.0    1613          18.0   
2   28.0          4          97.0        92.0    2288          17.0   
3   28.0          4          98.0        80.0    2164          15.0   
4   29.0          4          68.0        49.0    1867          19.5   
5   32.0          4          71.0        65.0    1836          21.0   
6   32.0          4          83.0        61.0    2003          19.0   
7   33.0          4          91.0        53.0    1795          17.5   
8   33.0          4          91.0        53.0    1795          17.4   
9   36.0          4          79.0        58.0    1825          18.6   
10  43.1          4          90.0        48.0    1985          21.5   
11  37.3          4          91.0        69.0    2130          14.7   
12  46.6          4          86.0        65.0    2110          17.9   
13  39.1          4          79.0        58.0    1755          16.9   
14  44.0          4          97.0        52.0    2130          24.6   

    model_year  origin                             name  
0           70   japan                     datsun pl510  
1           71   japan                      datsun 1200  
2           72   japan                  datsun 510 (sw)  
3           72     usa                  dodge colt (sw)  
4           73  europe                         fiat 128  
5           74   japan              toyota corolla 1200  
6           74   japan                       datsun 710  
7           75   japan                 honda civic cvcc  
8           76   japan                      honda civic  
9           77  europe                    renault 5 gtl  
10          78  europe  volkswagen rabbit custom diesel  
11          79  europe               fiat strada custom  
12          80   japan                        mazda glc  
13          81   japan                   toyota starlet  
14          82  europe                        vw pickup

In [30]:
result[['origin', 'name', 'weight','cylinders', 'mpg']]

origin                             name  weight  cylinders   mpg
0    japan                     datsun pl510    2130          4  27.0
1    japan                      datsun 1200    1613          4  35.0
2    japan                  datsun 510 (sw)    2288          4  28.0
3      usa                  dodge colt (sw)    2164          4  28.0
4   europe                         fiat 128    1867          4  29.0
5    japan              toyota corolla 1200    1836          4  32.0
6    japan                       datsun 710    2003          4  32.0
7    japan                 honda civic cvcc    1795          4  33.0
8    japan                      honda civic    1795          4  33.0
9   europe                    renault 5 gtl    1825          4  36.0
10  europe  volkswagen rabbit custom diesel    1985          4  43.1
11  europe               fiat strada custom    2130          4  37.3
12   japan                        mazda glc    2110          4  46.6
13   japan                   toyota starlet    1755          4  39.1
14  europe                        vw pickup    2130          4  44.0

In [31]:
sorted_result = result[['origin', 'name', 'weight','cylinders', 'mpg']].sort_values(by='mpg', ascending=False)
sorted_result

origin                             name  weight  cylinders   mpg
12   japan                        mazda glc    2110          4  46.6
14  europe                        vw pickup    2130          4  44.0
10  europe  volkswagen rabbit custom diesel    1985          4  43.1
13   japan                   toyota starlet    1755          4  39.1
11  europe               fiat strada custom    2130          4  37.3
9   europe                    renault 5 gtl    1825          4  36.0
1    japan                      datsun 1200    1613          4  35.0
7    japan                 honda civic cvcc    1795          4  33.0
8    japan                      honda civic    1795          4  33.0
5    japan              toyota corolla 1200    1836          4  32.0
6    japan                       datsun 710    2003          4  32.0
4   europe                         fiat 128    1867          4  29.0
2    japan                  datsun 510 (sw)    2288          4  28.0
3      usa                  dodge colt (sw)    2164          4  28.0
0    japan                     datsun pl510    2130          4  27.0

In [32]:
sorted_result.head(5)

origin                             name  weight  cylinders   mpg
12   japan                        mazda glc    2110          4  46.6
14  europe                        vw pickup    2130          4  44.0
10  europe  volkswagen rabbit custom diesel    1985          4  43.1
13   japan                   toyota starlet    1755          4  39.1
11  europe               fiat strada custom    2130          4  37.3

In [33]:
sorted_result.tail(5)

origin             name  weight  cylinders   mpg
6   japan       datsun 710    2003          4  32.0
4  europe         fiat 128    1867          4  29.0
2   japan  datsun 510 (sw)    2288          4  28.0
3     usa  dodge colt (sw)    2164          4  28.0
0   japan     datsun pl510    2130          4  27.0

In [35]:
mpg['origin'].unique()

array(['usa', 'japan', 'europe'], dtype=object)

In [37]:
mpg_comparison = mpg.groupby('origin')['mpg'].mean().reset_index()
mpg_comparison

origin        mpg
0  europe  27.891429
1   japan  30.450633
2     usa  20.083534

## 결과해석

- 여기서 mpg값이 가장 높은 차량은 46.6의 일본의 mazda glc이다
- mpg가 가장 높은 차량은 대부분 4기통 엔진이다
- 제조사별 mpg비교를 했을때 일본 > 유럽 > 미국 순으로 mpg값이 좋았다